tests/lightning/test_lightning_trainer.py

```diff
@@ -0,0 +1,346 @@
+        trainer = self.get_lightning_trainer(
+            max_updates=5,
+            max_epochs=None,
+            grad_clipping_config=self.grad_clipping_config,
+            callback=self,
+        )
+        self._prepare_trainer(trainer)
+        trainer.trainer.fit(trainer.model, trainer.data_module.train_loader)
+
+    def test_lr_schedule_with_or_without_is_different(self):
+        # note, be aware some of the logic also is in the SimpleLightningModel
+        trainer1 = self.get_lightning_trainer(max_updates=8, lr_scheduler=True)
+        self._prepare_trainer(trainer1)
+        trainer1.trainer.fit(trainer1.model, trainer1.data_module.train_loader)
+
+        trainer2 = self.get_lightning_trainer(max_updates=8)
+        self._prepare_trainer(trainer2)
+        trainer2.trainer.fit(trainer2.model, trainer2.data_module.train_loader)
+
+        last_model_param1 = list(trainer1.model.parameters())[-1]
+        last_model_param2 = list(trainer2.model.parameters())[-1]
+        self.assertFalse(torch.allclose(last_model_param1, last_model_param2))
+
+    def test_lr_schedule_compared_to_mmf_is_same(self):
+        trainer_config = self.get_trainer_config()
+        mmf_trainer = self.get_mmf_trainer(
+            max_updates=8, max_epochs=None, scheduler_config=trainer_config.scheduler
+        )
+        mmf_trainer.training_loop()
+
+        trainer = self.get_lightning_trainer(max_updates=8, lr_scheduler=True)
+        self._prepare_trainer(trainer)
+        trainer.trainer.fit(trainer.model, trainer.data_module.train_loader)
+
+        mmf_trainer.model.to(trainer.model.device)
+        last_model_param1 = list(mmf_trainer.model.parameters())[-1]
+        last_model_param2 = list(trainer.model.parameters())[-1]
+        self.assertTrue(torch.allclose(last_model_param1, last_model_param2))
+
+    def test_grad_accumulate(self):
+        trainer1 = self.get_lightning_trainer(
+            update_frequency=2, callback=self, max_updates=2, batch_size=3
+        )
+        self._prepare_trainer(trainer1)
+        trainer1.trainer.fit(trainer1.model, trainer1.data_module.train_loader)
+
+        trainer2 = self.get_lightning_trainer(
+            update_frequency=1, callback=self, max_updates=2, batch_size=6
+        )
+        self._prepare_trainer(trainer2)
+        trainer2.trainer.fit(trainer2.model, trainer2.data_module.train_loader)
+
+        for param1, param2 in zip(
+            trainer1.model.parameters(), trainer2.model.parameters()
+        ):
+            self.assertTrue(torch.allclose(param1, param2))
+
+    def _prepare_trainer(self, trainer):
+        trainer.configure_device()
+        trainer._calculate_max_updates()
+        trainer._calculate_gradient_clip_val()
+        trainer._load_trainer()
+
+    def test_same_loss_computation(self):
+        # check to see the same losses between the two trainers
+        # under the same conditions
+        self._test_same_loss_computation = True
```

Comment:
Is there a better approach to do this, this adds dynamic dependency between the tests and make them dependent certain conditions. Won't it be better to add new callback just for the sake of testing to the trainer?